pandas/tests/indexes/multi/test_integrity.py

```diff
@@ -0,0 +1,403 @@
+    inner = pd.DatetimeIndex([x[1] for x in result])
+    tm.assert_index_equal(inner, aware)
+
+    # n_lev > n_lab
+    result = idx[:2].values
+
+    outer = pd.DatetimeIndex([x[0] for x in result])
+    tm.assert_index_equal(outer, naive[:2])
+
+    inner = pd.DatetimeIndex([x[1] for x in result])
+    tm.assert_index_equal(inner, aware[:2])
+
+
+def test_values_multiindex_periodindex():
+    # Test to ensure we hit the boxing / nobox part of MI.values
+    ints = np.arange(2007, 2012)
+    pidx = pd.PeriodIndex(ints, freq='D')
+
+    idx = pd.MultiIndex.from_arrays([ints, pidx])
+    result = idx.values
+
+    outer = pd.Int64Index([x[0] for x in result])
+    tm.assert_index_equal(outer, pd.Int64Index(ints))
+
+    inner = pd.PeriodIndex([x[1] for x in result])
+    tm.assert_index_equal(inner, pidx)
+
+    # n_lev > n_lab
+    result = idx[:2].values
+
+    outer = pd.Int64Index([x[0] for x in result])
+    tm.assert_index_equal(outer, pd.Int64Index(ints[:2]))
+
+    inner = pd.PeriodIndex([x[1] for x in result])
+    tm.assert_index_equal(inner, pidx[:2])
+
+
+def test_consistency():
+    # need to construct an overflow
+    major_axis = lrange(70000)
+    minor_axis = lrange(10)
+
+    major_labels = np.arange(70000)
+    minor_labels = np.repeat(lrange(10), 7000)
+
+    # the fact that is works means it's consistent
+    index = MultiIndex(levels=[major_axis, minor_axis],
+                       labels=[major_labels, minor_labels])
+
+    # inconsistent
+    major_labels = np.array([0, 0, 1, 1, 1, 2, 2, 3, 3])
+    minor_labels = np.array([0, 1, 0, 1, 1, 0, 1, 0, 1])
+    index = MultiIndex(levels=[major_axis, minor_axis],
+                       labels=[major_labels, minor_labels])
+
+    assert not index.is_unique
+
+
+def test_hash_collisions():
+    # non-smoke test that we don't get hash collisions
+
+    index = MultiIndex.from_product([np.arange(1000), np.arange(1000)],
+                                    names=['one', 'two'])
+    result = index.get_indexer(index.values)
+    tm.assert_numpy_array_equal(result, np.arange(
+        len(index), dtype='intp'))
+
+    for i in [0, 1, len(index) - 2, len(index) - 1]:
+        result = index.get_loc(index[i])
+        assert result == i
+
+
+def test_dims():
+    pass
+
+
+def take_invalid_kwargs():
+    vals = [['A', 'B'],
+            [pd.Timestamp('2011-01-01'), pd.Timestamp('2011-01-02')]]
+    idx = pd.MultiIndex.from_product(vals, names=['str', 'dt'])
+    indices = [1, 2]
+
+    msg = r"take\(\) got an unexpected keyword argument 'foo'"
+    tm.assert_raises_regex(TypeError, msg, idx.take,
+                           indices, foo=2)
+
+    msg = "the 'out' parameter is not supported"
+    tm.assert_raises_regex(ValueError, msg, idx.take,
+                           indices, out=indices)
+
+    msg = "the 'mode' parameter is not supported"
+    tm.assert_raises_regex(ValueError, msg, idx.take,
+                           indices, mode='clip')
+
+
+def test_isna_behavior(idx):
```

Comment:
move to test_missing (anything na related is fair game for that)

Reply:
Done.